examples/rust/testdata/src/allocator.rs

```diff
@@ -0,0 +1,41 @@
+extern crate alloc;
+extern crate wee_alloc;
+
+use alloc::vec::Vec;
+use std::ffi::CString;
+use std::os::raw::c_char;
+
+/// Set the global allocator to the WebAssembly optimized one.
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+/// Allocates size bytes in linear memory and returns the offset where it starts.
+///
+/// In WebAssembly, the corresponding import will be:
+/// ```
+/// (import "xxx" "alloc" (func (param $size i32) (result i32)))
+///```
+#[no_mangle]
+pub extern "C" fn alloc(size: usize) -> *mut c_char {
+    // Allocate the amount of bytes we need, which is 1 more for the NUL suffix.
```

Comment:
```suggestion
    // Allocate the amount of bytes we need, which is 1 more than the requested size for the NUL suffix.
```

Reply:
Just want to confirm that we want to do the allocation within Rust, as opposed to passing in memory allocated in Go.

Reply:
it is a good question. the main concern is who controls allocation of the wasm memory.

Let's say we passed an arbitrary offset, like 8. When the rust-defined wasm initializes, it might have written to that area already (maybe some static data). We could choose a high number instead like 4000, but that also might collide with data allocated at runtime by rust-generated wasm.

The main thing we buy with using rust's allocator is that the responsibility of "picking offsets so that they don't collide" is moved there. I think this is a general win regardless of rust vs something else like TinyGo, as the problem would be the same. Basically, if the internal allocation impl isn't aware in any way of allocations (because we picked randomly outside with Go), it can easily collide with it.

The flip side is not all wasm is defined by a runtime that includes garbage collection. Ex if defining wasm with the text format, you have to manually track offsets or make a wasm library that does it and call that to manage allocation. Until things consistently use alloc/dealloc functions this sort of "library bias" will exist, and probably for a while considering garbage collection is "phase 2" https://github.com/WebAssembly/proposals#phase-2---proposed-spec-text-available-cg--wg